src/analysis/retail/shaman/elemental/modules/talents/Stormkeeper.tsx

```diff
@@ -1,37 +1,264 @@
+      Events.cast.by(SELECTED_PLAYER).spell(TALENTS.STORMKEEPER_1_ELEMENTAL_TALENT),
+      this.onSKCast,
+    );
     this.addEventListener(
-      Events.damage.by(SELECTED_PLAYER).spell(AFFECTED_ABILITIES),
+      Events.removebuff.by(SELECTED_PLAYER).spell(TALENTS.STORMKEEPER_1_ELEMENTAL_TALENT),
+      this.onSKFalloff,
+    );
+    this.addEventListener(
+      Events.damage.by(SELECTED_PLAYER).spell(SK_DAMAGE_AFFECTED_ABILITIES),
       this.onSKDamage,
     );
   }
 
+  onSKCast(cast: CastEvent) {
+    this.lastSKHardcast = cast;
+  }
+
+  onSKFalloff(event: RemoveBuffEvent) {
+    if (!this.activeWindow) {
+      return;
+    }
+
+    this.activeWindow.timeline.end = event.timestamp;
+
+    this.recordCooldown(this.activeWindow);
+    this.lastSKHardcast = null;
+    this.activeWindow = null;
+  }
+
+  onEventDuringSK(event: CastEvent) {
+    /* Don't include events that did not happen between hardcasts and the
+    SK buff falling off. */
+    if (!this.lastSKHardcast) {
+      return;
+    }
+
+    if (!this.activeWindow && WINDOW_START_SPELLS.includes(event.ability.guid)) {
+      this.activeWindow = {
+        event: this.lastSKHardcast,
+        maelstromOnCast: this.maelstromTracker.current,
+        flameshockDurationOnCast: this.enemies.getLongestDurationRemaining(
+          SPELLS.FLAME_SHOCK.id,
+          FLAMESHOCK_BASE_DURATION,
+          FLAMESHOCK_BASE_DURATION * (1 + PANDEMIC_WINDOW),
+          event.timestamp,
+        ),
+        electrifiedShocksDurationOnCast: this.enemies.getLongestDurationRemaining(
+          SPELLS.ELECTRIFIED_SHOCKS_DEBUFF.id,
+          ELECTRIFIED_SHOCKS_DURATION,
+          ELECTRIFIED_SHOCKS_DURATION,
+          event.timestamp,
+        ),
+        sopOnCast: this.selectedCombatant.hasBuff(SPELLS.SURGE_OF_POWER_BUFF.id),
+        moteOnCast: this.selectedCombatant.hasBuff(SPELLS.MASTER_OF_THE_ELEMENTS_BUFF.id),
+        timeline: {
+          start: this.lastSKHardcast.timestamp,
+          end: -1,
+          events: [this.lastSKHardcast],
+          performance: QualitativePerformance.Perfect,
+        },
+        firstRotationCast: event,
+      };
+
+      /* If the user started the window with a spender, then the maelstrom
+      has already been consumed */
+      if (event.ability.guid === this.stSpender.id) {
+        const resourceCost = event.resourceCost ? event.resourceCost : {};
+        this.activeWindow.maelstromOnCast += resourceCost[RESOURCE_TYPES.MAELSTROM.id] || 0;
+      }
+    }
+
+    if (this.activeWindow) {
+      const isPrepull = this.activeWindow.event.timestamp < this.owner.fight.start_time;
+      const spenderNotAlreadyCast = !this.activeWindow.timeline.events
+        .map((e) => e.ability.guid)
+        .includes(this.stSpender.id);
+
+      if (
+        event.ability.guid === this.stSpender.id &&
+        !this.selectedCombatant.hasBuff(
+          SPELLS.MASTER_OF_THE_ELEMENTS_BUFF.id,
+          event.timestamp,
+          100,
```

Comment:
[Do] I added a constant for this so we can be consistent about how we use it across modules.  It's `ON_CAST_BUFF_REMOVAL_GRACE_MS`.  Also, that way it links back to a comment about why it's there.